### PR TITLE
Add MATLAB support

### LINKED ISSUE #78: Add MATLAB support
Implement MATLAB codegen.

### PR DESCRIPTION
**Pull request by naegelejd**
Adds Matlab support for the yardl binary format.

Closes #78.

### CODE REVIEW

**Comment by naegelejd** [comment]
@johnstairs When you are ready, you can follow the [additions to `README.md`](https://github.com/microsoft/yardl/pull/147/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) to get started with Matlab.

On my Windows PC, where I run my devcontainer in WSL, my **WSL** shell profile `$HOME/.profile` looks like this:

```bash
export MATLAB_LICENSE_FILE=/mnt/c/Users/username/Downloads/Network.lic.com
```

Be sure to restart all instances of VS Code before building the devcontainer.

**Comment by johnstairs** [comment]
Should we have have the sandbox model generate MATLAB too? In the past, I've found that model to be a useful testbed for isolating problems

File: docs/index.md [hunk 1]
```diff
@@ -15,13 +15,16 @@ hero:
     - theme: brand
       text: Get Started With C++
       link: /cpp/installation
+    - theme: brand
+      text: Get Started With Matlab
```
Comment (johnstairs) [inline docs/index.md:19]: In docs and error error messages, should be referring to it as "MATLAB"?

**Review by johnstairs — COMMENTED** [review]
If you delete a top-level type (like a record or a protocol), we should make sure we delete the associated generated files.

**Review by johnstairs — COMMENTED** [review]
```yaml
MyType: !record
  fields:
    a: int->string
```

If I create two instances and call `a.eq(b)`, I get 
```
Error using ==
Concatenating 'dictionary' objects is not supported because 'dictionary' objects can only be scalar.
```

File: tooling/internal/matlab/types/types.go [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+		common.WriteBlockBody(w, func() {
+			w.WriteStringln("sz = [sz, sz];")
+		})
+		w.WriteStringln("z = reshape(repelem(elem, prod(sz)), sz);")
+	})
+}
+
+func typeEqualityExpression(t dsl.Type, a, b string) string {
+	if hasSimpleEquality(t) {
+		return fmt.Sprintf("all([%s] == [%s])", a, b)
```
Comment (johnstairs) [inline tooling/internal/matlab/types/types.go:397]: I'm wondering if this is the most efficient way of doing this, if a and b are (supposed to be), simple integers.

**Review by johnstairs — COMMENTED** [review]
```
MyType: !record
  fields:
    a: int*
```
If I create two instances of this type where the `a` fields have different lengths, `eq` raises an error:

```
Arrays have incompatible sizes for this operation.

Error in == (line 20)
        all([self.a] == [other.a]);
 ```

**Comment by johnstairs** [comment]
I don't know if we can customize a class' string representation, but say the default for union is not that helpful:
```
  Int32OrString with no properties.
  ```

File: docs/matlab/language.md [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+    recB: RecordB # But this is fine.
+
+RecordB: !record
+  fields:
+    c: int
+```
+
+Note that Yardl does not support type inheritance.
+
+The generated class constructors take arguments for each field and
```
Comment (johnstairs) [inline docs/matlab/language.md:129]: I would add a quick stample here showing the use of keyword args.

File: tooling/internal/python/static_files/_binary.py [hunk 4]
```diff
@@ -408,7 +408,7 @@ def write(self, stream: CodedOutputStream, value: Int16) -> None:
                     f"Value {value} is outside the range of a signed 16-bit integer"
                 )
         elif not isinstance(value, cast(type, np.int16)):
-            raise ValueError(f"Value in not an signed 16-bit integer: {value}")
+            raise ValueError(f"Value is not an signed 16-bit integer: {value}")
```
Comment (johnstairs) [inline tooling/internal/python/static_files/_binary.py:411]: Can you please fix my other typo while you're at it? 😆 "not a signed"

File: tooling/internal/matlab/static_files/+binary/FixedVectorSerializer.m [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+classdef FixedVectorSerializer < yardl.binary.VectorSerializerBase
+    properties
+        length_;
+    end
+
+    methods (Access=protected)
+        function handle_write_count_(self, ~, count)
+            if count ~= self.length_
+                throw(yardl.ValueError("Expected an array of length %d, got %d", self.length_, count));
```
Comment (johnstairs) [inline tooling/internal/matlab/static_files/+binary/FixedVectorSerializer.m:12]: A future improvement would be to provide a path with these errors, so the user has a better idea of what value caused the error. Applies to Python as well.

File: docs/matlab/language.md [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+
+```yaml
+Fruits: !enum
+  values:
+    - apple
+    - banana
+    - pear
+```
+
+Enums are generated as custom Matlab class definitions, *not* using Matlab's `enumeration` support, which doesn't allow integer values that are not explicitly defined in the enum definition.
```
Comment (johnstairs) [inline docs/matlab/language.md:310]: Did you consider using Matlab's enums, and retuning integers if the read value is outside the defined values?

Comment (naegelejd) [reply docs/matlab/language.md:310]: Yes, this is possible, however there is no good way to construct an enumeration from an integer outside the defined values. This means, for example, you can't mix an enumeration and equivalent integer in MATLAB arrays, which breaks yardl support for flags/enums in streams, vectors, and arrays:

```matlab
>> fs = [Fruits.APPLE, uint64(7)]
Error using Fruits
Cannot find a member of the 'Fruits' enumeration class that corresponds to each element of the given input
argument.
```

File: tooling/internal/matlab/static_files/+binary/FixedVectorSerializer.m (hunk 5 above)
Comment (naegelejd) [reply tooling/internal/matlab/static_files/+binary/FixedVectorSerializer.m:12]: Strongly agree, especially for records with many and/or nested fields. I vote to open a feature request issue after this PR.

File: tooling/internal/matlab/types/types.go (hunk 2 above)
Comment (naegelejd) [reply tooling/internal/matlab/types/types.go:397]: Fixed.

File: tooling/internal/matlab/types/types.go [hunk 7]
```diff
@@ -0,0 +1,1099 @@
...
+			w.WriteStringln("")
+			w.WriteStringln("function ne = ne(self, other)")
+			common.WriteBlockBody(w, func() {
+				w.WriteStringln("ne = ~self.eq(other);")
+			})
+			w.WriteStringln("")
+			w.WriteStringln("function t = tag(self)")
+			common.WriteBlockBody(w, func() {
+				fmt.Fprintf(w, "tags_ = [%s];\n", strings.Join(tags, ", "))
+				w.WriteStringln("t = tags_(self.index_);")
```
Comment (johnstairs) [inline tooling/internal/matlab/types/types.go:137]: Can we easily add a string representation of the value?

**Review by johnstairs — APPROVED** [review]
This is great, Joe!

File: tooling/internal/matlab/types/types.go (hunk 7 above)
Comment (naegelejd) [reply tooling/internal/matlab/types/types.go:137]: My last change added both the Union `tag` *and* the `value` to all Union string representations. When MATLAB displays an object property that is not a built-in type (e.g. all yardl-generated classes), it summarizes the property as a `[1x1 <classname>]`.

The following code
```matlab
i = 42;
disp(i);
u = test_model.AliasedIntOrSimpleRecord.Int32(i);
disp(u);
```
prints:
```
    42

AliasedIntOrSimpleRecord.Int32

    value: 42
````

and the following code
```matlab
sr = test_model.SimpleRecord(x=1, y=2, z=3);
disp(sr);
u = test_model.AliasedIntOrSimpleRecord.SimpleRecord(sr);
disp(u);
```
prints:
```
  SimpleRecord with properties:

    x: 1
    y: 2
    z: 3

AliasedIntOrSimpleRecord.SimpleRecord

    value: [1x1 test_model.SimpleRecord]
```

**Pull request merged**